src/jit/optimizer.cpp

```diff
@@ -1574,177 +1586,149 @@ void Compiler::optFindNaturalLoops()
-                 *
-                 * The condition we have to check for is
-                 *  1. ENTRY must have at least one predecessor inside the loop. Since we know that that block is
-                 *     reachable, it can only be reached through ENTRY, therefore we have a way back to ENTRY
-                 *
-                 *  2. If we have a GOTO (BBJ_ALWAYS) outside of the loop and that block dominates the
-                 *     loop bottom then we cannot iterate
-                 *
-                 * NOTE that this doesn't entirely satisfy condition 1. since "break" statements are not
-                 * part of the loop nodes (as per definition they are loop exits executed only once),
-                 * but we have no choice but to include them because we consider all blocks within TOP-BOTTOM */
-
-                for (loopBlock = top; loopBlock != bottom; loopBlock = loopBlock->bbNext)
-                {
-                    switch (loopBlock->bbJumpKind)
+                };
+                // Also we'll sometimes want to get a block's position using its predecessor's
+                // number if it is a new block, so define a helper for that.
+                auto positionNum = [oldBlockMaxNum](BasicBlock* block) -> unsigned int {
+                    if (block->bbNum > oldBlockMaxNum)
                     {
-                        case BBJ_ALWAYS:
-                        case BBJ_THROW:
-                        case BBJ_RETURN:
-                            if (fgDominate(loopBlock, bottom))
-                            {
-                                goto NO_LOOP;
-                            }
-                        default:
-                            break;
+                        // This must be a block we inserted to connect fall-through after moving blocks.
+                        // To determine if it's in the loop or not, use the number of its unique predecessor
+                        // block.
+                        assert(block->bbPreds->flBlock == block->bbPrev);
+                        assert(block->bbPreds->flNext == nullptr);
+                        return block->bbPrev->bbNum;
                     }
-                }
+                    return block->bbNum;
+                };
 
-                bool canIterateLoop = false;
+                // Seed the loop block set and worklist with the entry block.
+                LoopBlockSet              loopBlocks(this, oldBlockMaxNum, entry->bbNum);
+                jitstd::list<BasicBlock*> worklist(getAllocator());
+                worklist.push_back(entry);
 
-                for (predEntry = entry->bbPreds; predEntry; predEntry = predEntry->flNext)
+                while (!worklist.empty())
                 {
-                    if (predEntry->flBlock->bbNum >= top->bbNum && predEntry->flBlock->bbNum <= bottom->bbNum)
+                    BasicBlock* block = worklist.back();
+                    worklist.pop_back();
```

Comment:
It can; good catch, thanks.  Updated, but rather than using `positionNum` (since `fgDominate` takes the blocks), I just defer the check until processing the predecessor.